test/integration/header_prefix_integration_test.cc

```diff
@@ -0,0 +1,65 @@
+#include "test/integration/http_protocol_integration.h"
+#include "test/integration/server.h"
+
+#include "gtest/gtest.h"
+
+namespace Envoy {
+
+// Unfortunately in the Envoy test suite, the headers singleton is initialized
+// well before server start-up, so by the time the server has parsed the
+// bootstrap proto it's too late to set it.
+//
+// Instead, set the value early and regression test the bootstrap proto's validation of prefix
+// injection.
+
+static const char* custom_prefix_ = "x-custom";
+
+class HeaderPrefixIntegrationTest : public HttpProtocolIntegrationTest {
+public:
+  static void SetUpTestSuite() {
+    ThreadSafeSingleton<Http::PrefixValue>::get().setPrefix(custom_prefix_);
+  }
+  void TearDown() {
+    // Because server.cc calls setPrefix using bootstrap, make sure to setPrefix with something
```

Comment:
Sorry I'm confused here. Why is this necessary? Would it be better to just store an `std::string` inside the prefix object instead of a reference to an external string?

Reply:
I can, and that'd work fine for our use case, but from a thread safety perspective, saying "the underlying memory has to outlast the setPrefix call" makes sense where having a mutex "guarding" the underlying memory of the string which could be swapped out by another setPrefix call is really not clean.

Reply:
Sorry I don't follow? What's wrong with just copying the prefix internal to the object?

Reply:
So I think there's a paradigm in C++ which is "this points to a static variable, which will be around" and the static data pointed to can change.

If we store the memory internally it's IMO weird, because the mutex no longer really works. I guess with the guards around "no writes after reads" it's safe but it feels really icky.

I'll make the icky-to-me change since the RELEASE_ASSERT technically makes my concern more style than anything else.